include/nccl_ofi_mr.h

```diff
@@ -204,6 +204,39 @@ int nccl_ofi_mr_cache_insert_entry(nccl_ofi_mr_cache_t *cache, nccl_ofi_mr_ckey_
  */
 int nccl_ofi_mr_cache_del_entry(nccl_ofi_mr_cache_t *cache, void *handle);
 
+/**
+ * Generate a duplicated key with start and len aligned to cache's
+ * system_page_size. This is for callers of nccl_ofi_mr_cache_insert_entry()
+ * to register memory regions which fit the lookup page start and sizes.
+ * @return new nccl_ofi_mr_ckey_t, with base and len aligned.
+ */
+static inline nccl_ofi_mr_ckey_t nccl_ofi_mr_ckey_mk_aligned_dup(nccl_ofi_mr_ckey_ref ckey, nccl_ofi_mr_cache_t *cache)
+{
+	uintptr_t base_mask, page_base;
+	size_t size, aligned_size;
+	base_mask = -(cache->system_page_size);
+	page_base = nccl_ofi_mr_ckey_baseaddr(ckey) & base_mask; /* start of page of data */
+	size = nccl_ofi_mr_ckey_len(ckey);
+	aligned_size = (size & base_mask) /* higher bits */
+		       + ((size & ~base_mask) ? cache->system_page_size : 0); /* one extra page */
```

Comment:
cache->system_page_size should be the MIN of the reported system page size and 4 KiB, because system page size isn't reflective of the GDR mappings, and we need to not mess that up.  Remember, we're not actually trying to map a whole page (there's better ways to do that).  We're just trying to find an interval that makes an array-based cache manageable.